File: packages/python/elizaos/features/advanced_capabilities/services/follow_up.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.types import Service, ServiceType

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime


@dataclass
class FollowUpTask:
    entity_id: UUID
    reason: str
    message: str | None = None
    priority: str = "medium"
    scheduled_at: str = ""
    metadata: dict[str, str | int | float | bool] = field(default_factory=dict)


@dataclass
class FollowUpSuggestion:
    entity_id: UUID
    entity_name: str
    days_since_last_contact: int
    relationship_strength: float
    suggested_reason: str


class FollowUpService(Service):
    name = "follow_up"
    service_type = ServiceType.FOLLOW_UP

    @property
    def capability_description(self) -> str:
        return "Follow-up scheduling and reminder management service"

    def __init__(self) -> None:
        self._follow_ups: dict[UUID, FollowUpTask] = {}
        self._runtime: IAgentRuntime | None = None

    @classmethod
    async def start(cls, runtime: IAgentRuntime) -> FollowUpService:
        service = cls()
        service._runtime = runtime
        runtime.logger.info(
            "Follow-up service started",
            src="service:follow_up",
            agentId=str(runtime.agent_id),
        )
        return service

    async def stop(self) -> None:
        if self._runtime:
            self._runtime.logger.info(
                "Follow-up service stopped",
                src="service:follow_up",
                agentId=str(self._runtime.agent_id),
            )
        self._follow_ups.clear()
        self._runtime = None

    async def schedule_follow_up(
        self,
        entity_id: UUID,
        scheduled_at: datetime,
        reason: str,
        priority: str = "medium",
        message: str | None = None,
    ) -> FollowUpTask:
        task = FollowUpTask(
            entity_id=entity_id,
            reason=reason,
            message=message,
            priority=priority,
            scheduled_at=scheduled_at.isoformat(),
        )

        self._follow_ups[entity_id] = task

        if self._runtime:
            self._runtime.logger.info(
                f"Scheduled follow-up with {entity_id}",
                src="service:follow_up",
                scheduled_at=task.scheduled_at,
            )

        return task

    async def get_follow_up(self, entity_id: UUID) -> FollowUpTask | None:
        return self._follow_ups.get(entity_id)

    async def cancel_follow_up(self, entity_id: UUID) -> bool:
        if entity_id in self._follow_ups:
            del self._follow_ups[entity_id]
            return True
        return False

    async def get_upcoming_follow_ups(
        self,
        days_ahead: int = 7,
        include_overdue: bool = True,
    ) -> list[FollowUpTask]:
        now = datetime.now(UTC)
        results: list[FollowUpTask] = []

        for task in self._follow_ups.values():
            scheduled = datetime.fromisoformat(task.scheduled_at.replace("Z", "+00:00"))
            days_until = (scheduled - now).days

            if include_overdue and days_until < 0 or 0 <= days_until <= days_ahead:
                results.append(task)

        results.sort(key=lambda t: t.scheduled_at)
        return results

    async def get_overdue_follow_ups(self) -> list[FollowUpTask]:
        now = datetime.now(UTC)
        results: list[FollowUpTask] = []

        for task in self._follow_ups.values():
            scheduled = datetime.fromisoformat(task.scheduled_at.replace("Z", "+00:00"))
            if scheduled < now:
                results.append(task)

        return results

    async def get_follow_up_suggestions(
        self,
        max_suggestions: int = 5,
    ) -> list[FollowUpSuggestion]:
        return []

    async def complete_follow_up(self, entity_id: UUID) -> bool:
        return await self.cancel_follow_up(entity_id)

File: packages/rust/src/features/basic_capabilities/bootstrap/services/follow_up.rs
//! Follow-up service implementation.

use async_trait::async_trait;
use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;

use super::{Service, ServiceType};

/// Follow-up task data.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FollowUpTask {
    pub entity_id: Uuid,
    pub reason: String,
    pub message: Option<String>,
    pub priority: String,
    pub scheduled_at: DateTime<Utc>,
}

/// Follow-up suggestion.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FollowUpSuggestion {
    pub entity_id: Uuid,
    pub entity_name: String,
    pub days_since_last_contact: i64,
    pub relationship_strength: f64,
    pub suggested_reason: String,
}

/// Service for managing follow-up reminders.
pub struct FollowUpService {
    follow_ups: HashMap<Uuid, FollowUpTask>,
    runtime: Option<Arc<dyn IAgentRuntime>>,
}

impl FollowUpService {
    /// Create a new follow-up service.
    pub fn new() -> Self {
        Self {
            follow_ups: HashMap::new(),
            runtime: None,
        }
    }

    /// Schedule a follow-up.
    pub fn schedule_follow_up(
        &mut self,
        entity_id: Uuid,
        scheduled_at: DateTime<Utc>,
        reason: String,
        priority: String,
        message: Option<String>,
    ) -> FollowUpTask {
        let task = FollowUpTask {
            entity_id,
            reason,
            message,
            priority,
            scheduled_at,
        };

        if let Some(runtime) = &self.runtime {
            runtime.log_info(
                "service:follow_up",
                &format!("Scheduled follow-up with {}", entity_id),
            );
        }

        self.follow_ups.insert(entity_id, task.clone());
        task
    }

    /// Get a follow-up by entity ID.
    pub fn get_follow_up(&self, entity_id: Uuid) -> Option<&FollowUpTask> {
        self.follow_ups.get(&entity_id)
    }

    /// Cancel a follow-up.
    pub fn cancel_follow_up(&mut self, entity_id: Uuid) -> bool {
        self.follow_ups.remove(&entity_id).is_some()
    }

    /// Get upcoming follow-ups.
    pub fn get_upcoming_follow_ups(
        &self,
        days_ahead: i64,
        include_overdue: bool,
    ) -> Vec<&FollowUpTask> {
        let now = Utc::now();
        let future = now + Duration::days(days_ahead);

        self.follow_ups
            .values()
            .filter(|t| {
                if include_overdue && t.scheduled_at < now {
                    return true;
                }
                t.scheduled_at >= now && t.scheduled_at <= future
            })
            .collect()
    }

    /// Get overdue follow-ups.
    pub fn get_overdue_follow_ups(&self) -> Vec<&FollowUpTask> {
        let now = Utc::now();
        self.follow_ups
            .values()
            .filter(|t| t.scheduled_at < now)
            .collect()
    }

    /// Complete a follow-up.
    pub fn complete_follow_up(&mut self, entity_id: Uuid) -> bool {
        self.cancel_follow_up(entity_id)
    }
}

impl Default for FollowUpService {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Service for FollowUpService {
    fn name(&self) -> &'static str {
        "follow_up"
    }

    fn service_type(&self) -> ServiceType {
        ServiceType::Core
    }

    async fn start(&mut self, runtime: Arc<dyn IAgentRuntime>) -> PluginResult<()> {
        runtime.log_info("service:follow_up", "Follow-up service started");
        self.runtime = Some(runtime);
        Ok(())
    }

    async fn stop(&mut self) -> PluginResult<()> {
        if let Some(runtime) = &self.runtime {
            runtime.log_info("service:follow_up", "Follow-up service stopped");
        }
        self.follow_ups.clear();
        self.runtime = None;
        Ok(())
    }
}
